### PR TITLE
Make Typer#dyna.mkInvoke more accurate

### PR DESCRIPTION
**Pull request by howtonotwin**
When searching the context for the tree in question, make sure to
actually check that the tree under scrutiny matches the one we're
looking for. This means that the check for varargs won't give a false
positive if the context is too large. Fixes scala/bug#7420.

Adjust the "vararg not supported" error so it doesn't hide other errors.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -4184,22 +4179,26 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
           def findSelection(t: Tree): Option[(TermName, Tree)] = t match {
-            case Apply(fn, args) if hasStar(args) => DynamicVarArgUnsupported(tree, applyOp(args)) ; None
-            case Apply(fn, args) if matches(fn)   => Some((applyOp(args), fn))
-            case Assign(lhs, _) if matches(lhs)   => Some((nme.updateDynamic, lhs))
-            case _ if matches(t)                  => Some((nme.selectDynamic, t))
-            case _                                => (t.children flatMap findSelection).headOption
+            case Apply(fn, args) if matches(fn) =>
+              val op = if(args.exists(_.isInstanceOf[AssignOrNamedArg])) nme.applyDynamicNamed else nme.applyDynamic
+              // not supported: foo.bar(a1,..., an: _*)
+              val fn1 = if(treeInfo.isWildcardStarArgList(args)) DynamicVarArgUnsupported(fn, op) else fn
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:4185]: `fn1` is unused

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 2]
```diff
@@ -622,8 +622,10 @@ trait ContextErrors {
         NormalTypeError(tree, fun.tpe+" does not take parameters")
 
       // Dynamic
-      def DynamicVarArgUnsupported(tree: Tree, name: Name) =
-        issueNormalTypeError(tree, name+ " does not support passing a vararg parameter")
+      def DynamicVarArgUnsupported(tree: Tree, name: Name) = {
+        issueNormalTypeError(tree, name + " does not support passing a vararg parameter")
+        setError(tree)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:627]: The new errors (reported in the test case) are not helpful / confusing, I'd say let's remove this again.

**Review by lrytz — COMMENTED** [review]
Looks great, just two minor comments.

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala (hunk 2 above)
Comment (howtonotwin) [reply src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:627]: I just found it rather irregular/strange that the "varargs not supported" error masked this (IMO more important) one. Besides, there *are* two separate errors here: you can't use varargs and you are missing the method itself, and they may as well be given up front.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:627]: OK, you convinced me :)

**Review by lrytz — APPROVED** [review]
Thank you!

**Pull request merged**